### PR TITLE
feat: introduced functionality to store selected values on matching pages when user errors

### PR DESCRIPTION
**Pull request by gilesdrinkall**
# Description

-   To introduce functionality so that the values selected by users on the matching pages are retained even if they trigger an error by not assigning all of the stages at least once.

# Testing instructions

-   This functionality can be tested locally on the fdbt-dev repo by running the 'make' command.

# Type of change

-   [X] feat - A new feature
-   [ ] fix - A bug fix
-   [ ] docs - Documentation only changes
-   [ ] codestyle - Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc)
-   [ ] refactor - A code change that neither fixes a bug nor adds a feature
-   [ ] perf - A code change that improves performance
-   [ ] test - Adding missing tests or correcting existing tests
-   [ ] build - Changes that affect the build system or external dependencies (example scopes: gulp, broccoli, npm)
-   [ ] ci - Changes to our CI configuration files and scripts (example scopes: Travis, Circle, BrowserStack, SauceLabs)
-   [ ] chore - Other changes that don't modify src or test files
-   [ ] revert - Reverts a previous commit
-   [ ] content - Changes to content or copy

# Checklist:

-   [X] Able to run pr locally
-   [ ] Lighthouse performed (Accessibility 90+ minimum)
-   [X] Followed acceptance criteria
-   [X] My code follows the style guidelines of this project
-   [X] I have performed a self-review of my own code
-   [ ] I have made corresponding changes to the documentation if applicable
-   [X] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: src/components/MatchingList.tsx [hunk 1]
```diff
@@ -6,42 +6,66 @@ import { formatStopName } from '../utils';
...
+
+                const isSelected = selectedFareStages.some(selectedObject => {
+                    return selectedObject === currentValue;
+                });
+
+                if (isSelected) {
+                    selectValue = currentValue;
+                }
+
+                return '';
```
Comment (tejpowar) [inline src/components/MatchingList.tsx:33]: maybe we change this to return null

File: src/pages/api/matching.ts [hunk 2]
```diff
@@ -122,31 +97,32 @@ export default async (req: NextApiRequest, res: NextApiResponse): Promise<void>
...
             throw new Error('No service or userfarestages info found');
         }
-
         const service: BasicService = JSON.parse(req.body.service);
         const userFareStages: UserFareStages = JSON.parse(req.body.userfarestages);
-
         const matchingFareZones = getMatchingFareZonesFromForm(req);
-
+        delete req.body.service;
+        delete req.body.userfarestages;
```
Comment (tejpowar) [inline src/pages/api/matching.ts:110]: add a new line here and above so clearer

Comment (tejpowar) [reply src/pages/api/matching.ts:110]: worth adding this comment  
 // Deleting these keys from the object in order to faciliate looping through the fare stage values in the body

File: src/pages/matching.tsx [hunk 3]
```diff
@@ -100,6 +103,8 @@ export const getServerSideProps = async (ctx: NextPageContext): Promise<{ props:
         .map(atco => naptanInfo.find(s => s.atcoCode === atco))
         .filter((stop: Stop | undefined): stop is Stop => stop !== undefined);
 
+    console.log(matchingCookie);
```
Comment (tejpowar) [inline src/pages/matching.tsx:106]: unwanted console.log

File: src/components/MatchingList.tsx [hunk 4]
```diff
@@ -6,42 +6,66 @@ import { formatStopName } from '../utils';
...
-                        >
-                            <option value="">Not Applicable</option>
-
-                            {userFareStages.fareStages.map((stage: FareStage) => (
-                                <option key={stage.stageName} value={JSON.stringify({ stop, stage: stage.stageName })}>
-                                    {stage.stageName}
-                                </option>
-                            ))}
-                        </select>
+        {stops.map((stop, index) => {
```
Comment (RobCatton) [inline src/components/MatchingList.tsx:19]: Struggling to understand what's going on here - would be good to have a chat to grok it :)

File: src/components/MatchingList.tsx (hunk 1 above)
Comment (gilesdrinkall) [reply src/components/MatchingList.tsx:33]: Changed to return null.

File: src/pages/api/matching.ts (hunk 2 above)
Comment (gilesdrinkall) [reply src/pages/api/matching.ts:110]: Added both lines and comment.

File: src/pages/matching.tsx (hunk 3 above)
Comment (gilesdrinkall) [reply src/pages/matching.tsx:106]: Deleted console.log

**Pull request merged**